src/file_watcher.rs

```diff
@@ -0,0 +1,1434 @@
+        | (?:^|\s+) (?P<message_only>[^\=]+) \s* $
+        | (?:^|\s+) (?P<message>[^\=]+) \s+
+        | (?P<left>[^\s\=]+) = (?P<right>[^\=]+) (?:\s+|$)
+        ",
+    )?)
+});
+
+impl FileWatcher {
+    /// Create a new `FileWatcher`
+    ///
+    /// Warning: This function should only be called in applications and not
+    /// within libraries as it will clobber any set by the application.
+    ///
+    /// Warning: If you will be creating a `FileWatcher` and a `TelemetryLayer`,
+    /// you must create the `FileWatcher` before the `TelemetryLayer` as there is
+    /// a race condition in the thread runtime of `tracing` and the tokio runtime
+    /// of `Reqwest`. Swapping order of the two could lead to possible deadlocks.
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            client: None,
+            request: None,
+            lockfile_path: Path::new(&telemetry_config()?.fuelup_tmp)
+                .join(config()?.lockfile.clone()),
+        })
+    }
+
+    /// Start the `FileWatcher`
+    ///
+    /// This function forks and has the parent immediately return. The forked
+    /// off child then daemonises to poll for aged-out telemetry files, sending
+    /// them to InfluxDB, and exits once there are no more files to process.
+    pub fn start(&mut self) -> WatcherResult<()> {
+        self.start_with_helpers(&DefaultStartHelpers)
+    }
+
+    fn start_with_helpers(&mut self, helpers: &impl StartHelpers) -> WatcherResult<()> {
+        if var("FUELUP_NO_TELEMETRY").is_ok() {
+            // If telemetry is disabled, immediately return
+            return Ok(());
+        }
+
+        let logfile = &config().map_err(into_recoverable)?.logfile;
+
+        if STARTED.load(Ordering::Relaxed) {
+            return Ok(());
+        } else {
+            STARTED.store(true, Ordering::Relaxed);
+        }
+
+        match helpers.daemonise(logfile) {
+            Ok(Some(pid)) => {
+                // We are the parent, so record the PID then immediately return
+                PID.store(pid.as_raw(), Ordering::Relaxed);
+                return Ok(());
+            }
+            Err(e) => {
+                // Couldn't daemonise, so clear the PID before returning the error
+                STARTED.store(false, Ordering::Relaxed);
+                return Err(e);
+            }
+            Ok(None) => {
+                // We are the child process, so continue as the `FileWatcher`...
+                PID.store(getpid().as_raw(), Ordering::Relaxed);
+            }
+        }
+
+        // From here on, we are no longer the original process, so the caller should
+        // treat errors as fatal. This means that on error the process should exit
+        // immediately as there should not be two identical flows of execution
+        // from the caller's perspective.
+
+        // Cache the client and request so we don't recreate them each time
+        self.client = Some(Client::new());
```

Comment:
nit: consider specifying timeout with client - to prevent the process from hanging indefinitely on network issues

Reply:
All good here - by default `reqwest::blocking::Client` has a 30 second timeout.